registry/txt.go

```diff
@@ -17,143 +17,127 @@ limitations under the License.
-	txtLabelRegex  = regexp.MustCompile("^\"heritage=external-dns,external-dns/owner=(.+)\"")
-	txtLabelFormat = "\"heritage=external-dns,external-dns/owner=%s\""
+const (
+	heritageLabel  = "heritage"
+	heritageValue  = "external-dns"
+	labelKeyPrefix = "external-dns/"
+	txtRecordType  = "TXT"
 )
 
-// TXTRegistry implements registry interface with ownership implemented via associated TXT records
+// TXTRegistry stores labels as DNS TXT records in the wrapped Provider.
 type TXTRegistry struct {
 	provider provider.Provider
-	ownerID  string //refers to the owner id of the current instance
-	mapper   nameMapper
 }
 
-// NewTXTRegistry returns new TXTRegistry object
-func NewTXTRegistry(provider provider.Provider, txtPrefix, ownerID string) (*TXTRegistry, error) {
-	if ownerID == "" {
-		return nil, errors.New("owner id cannot be empty")
-	}
-
-	mapper := newPrefixNameMapper(txtPrefix)
-
-	return &TXTRegistry{
-		provider: provider,
-		ownerID:  ownerID,
-		mapper:   mapper,
-	}, nil
+// NewTXTRegistry returns a new DNS-based registry for the passed in Provider.
+func NewTXTRegistry(prov provider.Provider, _ string) (*TXTRegistry, error) {
+	return &TXTRegistry{provider: prov}, nil
 }
 
-// Records returns the current records from the registry excluding TXT Records
-// If TXT records was created previously to indicate ownership its corresponding value
-// will be added to the endpoints Labels map
-func (im *TXTRegistry) Records() ([]*endpoint.Endpoint, error) {
-	records, err := im.provider.Records()
+// Records reads DNS records from the nested provider and translates any TXT records containing
+// label information into labels on the returned Endpoint objects.
+func (p *TXTRegistry) Records() ([]*endpoint.Endpoint, error) {
+	// Read all records from the DNS provider.
+	records, err := p.provider.Records()
 	if err != nil {
 		return nil, err
 	}
 
-	endpoints := make([]*endpoint.Endpoint, 0)
-
-	ownerMap := map[string]string{}
+	// A temporary store that holds all found labels.
+	labels := map[string]map[string]string{}
 
-	for _, record := range records {
-		if record.RecordType != "TXT" {
-			endpoints = append(endpoints, record)
+	// Go through all TXT records and temporary collect and store label information.
+	for _, r := range records {
+		// Only consider records of type TXT.
+		if r.RecordType != txtRecordType {
 			continue
 		}
-		ownerID := im.extractOwnerID(record.Target)
-		if ownerID == "" {
-			//case when value of txt record cannot be identified
-			//record will not be removed as it will have empty owner
-			endpoints = append(endpoints, record)
+
+		// Parse the value of the TXT record into a label set (a key-value map).
+		labelMap := parseLabels(r.Target)
+
+		// If the TXT record indicates a heritage of ExternalDNS, then attach the remaining labels
+		// to the endpoint object. (More specific: store them to be attached in the following section.)
+		// If not, skip this record.
+		if labelMap[heritageLabel] != heritageValue {
```

Comment:
I believe this "TXT" record which does not have heritage label should be included in the endpoints list, because it is not meant to be deleted and should be regarded as other parties A, CNAME